Incorrect return type hint in `SymbolicReference.reference` leads to mypy errors

Since GitPython 3.1.45 was released, we've had `mypy` reporting pre-checkin errors for code that has been stable and functional for years:
```
error: "SymbolicReference" has no attribute "checkout"  [attr-defined]
error: "SymbolicReference" has no attribute "tracking_branch"  [attr-defined]
error: "SymbolicReference" has no attribute "tracking_branch"  [attr-defined]
```

The code failing `mypy` checks is doing things like getting the active branch, so we can switch back to it after doing other stuff:
```
    @contextmanager
    def checkout(self, branch: Branch) -> typing.Generator[None]:
        new_branch = self.api.create_head(branch.name)
        current = self.api.head.ref
        try:
            new_branch.checkout()
            yield
        finally:
            current.checkout()
```

... or ensuring that the active branch has a remote tracking branch set:
```
    tracking_branch = repo.head.reference.tracking_branch()
    if tracking_branch is None:
        raise GitConfigException(
            'Error: Upstream branch not set; use "git branch '
            '--set-upstream-to=<remote/branch>" to fix.'
        )
    if tracking_branch.remote_name == ".":
        raise GitConfigException(
            'Error: Upstream branch is local rather than remote; use "git '
            'branch --set-upstream-to=<remote/branch>" to fix.'
        )
``` 

The immediate cause of these new `mypy` errors appears to be [this commit](https://github.com/gitpython-developers/GitPython/commit/94151aa2ca9f16491a0cf2344b4daa8bf7b41d70#diff-72d42177f0e8535634711b5de0390d5bc81dbedab1fbac6547e5b9ab3b03eb9aL52), which added a new type hint for the return value of `SymbolicReference.reference`:
```
    @property
    def reference(self) -> "SymbolicReference":
        return self._get_reference()
```

In manual testing I've found that the `reference` method doesn't actually return a `SymbolicReference` object when the the symbolic reference is `HEAD`. Instead, a `Head` object is returned:
```
>>> import git
>>> git.__version__
'3.1.45'
>>> testrepo = git.repo.Repo.init(".")
>>> type(testrepo.head)
<class 'git.refs.head.HEAD'>
>>> type(testrepo.head.ref)
<class 'git.refs.head.Head'>
>>> testrepo.head.ref.tracking_branch()
<git.RemoteReference "refs/remotes/./master">
>>> testrepo.head.ref.checkout()
<git.Head "refs/heads/bar">
```

--- a/git/refs/symbolic.py
+++ b/git/refs/symbolic.py
@@ -40,6 +40,7 @@ if TYPE_CHECKING:
     from git.config import GitConfigParser
     from git.objects.commit import Actor
     from git.refs.log import RefLogEntry
+    from git.refs.reference import Reference
     from git.repo import Repo
 
 
@@ -404,7 +405,7 @@ class SymbolicReference:
     def object(self, object: Union[AnyGitObject, "SymbolicReference", str]) -> "SymbolicReference":
         return self.set_object(object)
 
-    def _get_reference(self) -> "SymbolicReference":
+    def _get_reference(self) -> "Reference":
         """
         :return:
             :class:`~git.refs.reference.Reference` object we point to
@@ -416,7 +417,7 @@ class SymbolicReference:
         sha, target_ref_path = self._get_ref_info(self.repo, self.path)
         if target_ref_path is None:
             raise TypeError("%s is a detached symbolic reference as it points to %r" % (self, sha))
-        return self.from_path(self.repo, target_ref_path)
+        return cast("Reference", self.from_path(self.repo, target_ref_path))
 
     def set_reference(
         self,
@@ -502,7 +503,7 @@ class SymbolicReference:
 
     # Aliased reference
     @property
-    def reference(self) -> "SymbolicReference":
+    def reference(self) -> "Reference":
         return self._get_reference()
 
     @reference.setter

